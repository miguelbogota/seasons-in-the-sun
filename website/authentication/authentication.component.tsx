import { TextField } from '@app/ui/text-field';
import { zodResolver } from '@hookform/resolvers/zod';
import { useForm } from 'react-hook-form';

import * as styles from './authentication.css';
import {
  type AppSigninCredentials,
  appSigninCredentialsSchema,
  type AppSignupCredentials,
  appSignupCredentialsSchema,
  useAuthentication,
} from './state';

export function Authentication() {
  const { isLoading, user, signout } = useAuthentication();

  return isLoading ? (
    'loading...'
  ) : (
    <div className={styles.root}>
      <pre>{JSON.stringify(user, null, 2)}</pre>

      {user ? (
        <>
          <button onClick={signout}>Sign out</button>
        </>
      ) : (
        <>
          <Signin />
          <br />
          <div>--------------------------------------------</div>
          <br />
          <Signup />
        </>
      )}
    </div>
  );
}

function Signin() {
  const { signin } = useAuthentication();
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<AppSigninCredentials>({
    criteriaMode: 'all',
    resolver: zodResolver(appSigninCredentialsSchema),
  });

  const onSubmit = handleSubmit(async (data) => {
    await signin(data).catch(console.error);
  });

  return (
    <div>
      <h2>Welcome Back</h2>
      <form onSubmit={onSubmit}>
        <TextField
          {...register('username')}
          label="Username"
          placeholder="Ex. johndoe"
          hint={errors.username?.message}
        />

        <TextField {...register('password')} label="Password" hint={errors.password?.message} />

        <button type="submit">Sign in</button>
      </form>
    </div>
  );
}

function Signup() {
  const { signup } = useAuthentication();
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<AppSignupCredentials>({
    criteriaMode: 'all',
    resolver: zodResolver(appSignupCredentialsSchema),
  });

  const onSubmit = handleSubmit(async (data) => {
    await signup(data).catch(console.error);
  });

  return (
    <div>
      <h2>Join</h2>
      <form onSubmit={onSubmit}>
        <TextField
          {...register('username')}
          label="Username"
          placeholder="Ex. johndoe"
          hint={errors.username?.message}
        />

        <TextField {...register('password')} label="Password" />
        {errors?.password?.types &&
          Object.entries(errors.password.types).map(([type, message]) =>
            Array.isArray(message) ? (
              message.map((message) => (
                <>
                  <p key={type}>{message}</p>
                  <br />
                </>
              ))
            ) : (
              <>
                <p key={type}>{message}</p>
                <br />
              </>
            ),
          )}

        <TextField
          {...register('confirmPassword')}
          label="Confirm Password"
          hint={errors.confirmPassword?.message}
        />

        <TextField
          {...register('email')}
          label="Email"
          placeholder="Ex. johndoe@email.com"
          hint={errors.email?.message}
        />

        <button type="submit">Sign up</button>
      </form>
    </div>
  );
}
