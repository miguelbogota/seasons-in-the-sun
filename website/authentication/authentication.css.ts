import { style } from '@vanilla-extract/css';

export const root = style({
  padding: '1rem',
  backgroundColor: '#ffffff',
  borderRadius: 3,
});
