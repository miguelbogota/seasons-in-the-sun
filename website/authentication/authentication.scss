@use './signin/signin' as signin;
@use './signup/signup' as signup;

@mixin apply {
  .authentication {
    background-color: var(--background-sheet);
    border-radius: 3px;

    &:has(.authenticated-menu) {
      background-color: unset;
    }

    .text-field {
      display: block;
      margin: 0.6rem 0;
    }

    .button {
      display: block;
      text-align: center;
      width: 100%;
      padding: 0.5rem;
      font-weight: 700;
      text-transform: uppercase;
      color: var(--white);
      background-color: var(--primary);
      border: none;
      border-radius: 4px;
      margin-bottom: 1rem;
      cursor: pointer;
      transition: background-color 0.2s ease-in-out;
    }

    h2 {
      text-align: center;
      font-size: var(--size-5);
      margin-bottom: var(--space-3);
    }

    & > .authenticated-menu {
      position: relative;

      & > .user-icon {
        display: flex;
        align-items: center;
        justify-content: center;
        background-color: var(--primary);
        border-radius: var(--radius-rounded);
        cursor: pointer;

        & > img {
          width: 4rem;
          height: 4rem;
          border-radius: 50%;
        }
      }

      & > .dropdown-menu {
        position: absolute;
        top: calc(100% + var(--space-1));
        right: 0;
        background-color: var(--background-sheet);
        border-radius: var(--radius-normal);
        box-shadow: 0 0 0.5rem 0 var(--shadow);
        padding: var(--space-3) var(--space-3) 0 var(--space-3);
        z-index: var(--modal);

        & > h4 {
          font-weight: var(--weight-bold);
          text-transform: uppercase;
          text-wrap: nowrap;
          margin-bottom: var(--space-1);
        }

        & > p {
          margin-bottom: var(--space-3);
        }
      }
    }

    @include signin.apply;
    @include signup.apply;
  }
}
