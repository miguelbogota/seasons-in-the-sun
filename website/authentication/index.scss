@use './authentication' as authentication;

@mixin apply {
  @include authentication.apply;
}
