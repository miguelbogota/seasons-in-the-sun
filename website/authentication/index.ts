export { Authentication } from './authentication.component';
export {
  type AppSigninCredentials,
  appSigninCredentialsSchema,
  type AppSignupCredentials,
  appSignupCredentialsSchema,
  type AppUser,
  useAuthentication,
} from './state';
