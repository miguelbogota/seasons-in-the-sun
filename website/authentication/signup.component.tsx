// import * as styles from './signup.css';

export function Signup() {
  return <div>Sign up</div>;
}
