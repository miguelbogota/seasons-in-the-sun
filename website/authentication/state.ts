import { gun, type UserSpace } from '@app/lib/gun';
import { useEffect } from 'react';
import { z } from 'zod';
import { create as createStore } from 'zustand';

/**
 * Authentication state hook.
 *
 * @param onAuthChanges Callback to be called when the user's authentication state changes.
 */
export function useAuthentication(onAuthChanges?: (user: AppUser | null) => void) {
  const { isLoading, user } = useAuthenticationState();

  // Subscribe to changes in the user's authentication state.
  useEffect(() => {
    const unsubscribe = gun?.()
      .user()
      .authChanges((u) => {
        onAuthChanges?.(u);
        useAuthenticationState.setState({ isLoading: false, user: u });
      });

    return unsubscribe;
  }, []);

  // Sign in function.
  const signin = async (credential: AppSigninCredentials) => {
    const { username, password } = credential;
    const user = await gun?.().user().auth(username, password);
    return user;
  };

  return {
    /** State to validate if any process is in progress. */
    isLoading,
    /** Current authenticated user (If null it means the user is not authenticated). */
    user,
    /** Helper to validate simpler the if the user is authenticated. */
    is: !!user,
    /** Sign in function. */
    signin: async (credential: AppSigninCredentials) => {
      const validation = await appSigninCredentialsSchema.safeParseAsync(credential);

      if (!validation.success) {
        throw validation.error.errors;
      }

      useAuthenticationState.setState({ isLoading: true });
      const user = await signin(credential).catch(() => {
        useAuthenticationState.setState({ isLoading: false });
      });
      return user;
    },
    /** Sign up function (If user signups successfully it also signs in). */
    signup: async (credential: AppSignupCredentials) => {
      const { username, password, confirmPassword: _, ...metaData } = credential;

      const validation = await appSignupCredentialsSchema.safeParseAsync(credential);

      if (!validation.success) {
        throw validation.error.errors;
      }

      useAuthenticationState.setState({ isLoading: true });
      const createdUser = await gun?.()
        .user()
        .create(username, password, {
          ...metaData,
          imageUrl: `https://api.dicebear.com/7.x/pixel-art/svg?seed=${username}`,
        })
        .catch(() => {
          useAuthenticationState.setState({ isLoading: false });
        });

      return createdUser;
    },
    /** Sign out function. */
    signout: async () => {
      useAuthenticationState.setState({ isLoading: true });
      await gun?.()
        .user()
        .leave()
        .catch(() => {
          useAuthenticationState.setState({ isLoading: false });
        });
    },
  };
}

/**
 * Authentication state.
 */
type AuthenticationState = {
  user: null | AppUser;
  isLoading: boolean;
};

/**
 * Authentication state store.
 */
const useAuthenticationState = createStore<AuthenticationState>(() => ({
  // The user is updated in the `authChanges` callback.
  user: null,
  isLoading: true,
}));

/**
 * App user type.
 */
export type AppUser = UserSpace['root'];

/**
 * Sign in credentials validation schema. No need to validate a lot the fields since it should be
 * already validated when the user was created
 */
export const appSigninCredentialsSchema = z.object({
  username: z
    .string({
      description: 'Username',
      required_error: 'Username is required',
    })
    .min(10, 'Username must be at least 10 characters')
    .max(30, 'Username must be at most 30 characters'),
  password: z
    .string({
      description: 'Password',
      required_error: 'Password is required',
    })
    .min(10, 'Password must be at least 10 characters')
    .max(30, 'Password must be at most 30 characters')
    .regex(/[a-z]/, 'Password must contain at least one lowercase letter')
    .regex(/[A-Z]/, 'Password must contain at least one uppercase letter')
    .regex(/[0-9]/, 'Password must contain at least one number')
    .regex(/[^a-zA-Z0-9]/, 'Password must contain at least one special character'),
});

/**
 * Sign in credentials type.
 */
export type AppSigninCredentials = z.infer<typeof appSigninCredentialsSchema>;

export const appSignupCredentialsSchema = z
  .object({
    username: z
      .string({
        description: 'Username',
        required_error: 'Username is required',
      })
      .min(10, 'Username must be at least 10 characters')
      .max(30, 'Username must be at most 30 characters'),
    password: z
      .string({
        description: 'Password',
        required_error: 'Password is required',
      })
      .min(10, 'Password must be at least 10 characters')
      .max(30, 'Password must be at most 30 characters')
      .regex(/[a-z]/, 'Password must contain at least one lowercase letter')
      .regex(/[A-Z]/, 'Password must contain at least one uppercase letter')
      .regex(/[0-9]/, 'Password must contain at least one number')
      .regex(/[^a-zA-Z0-9]/, 'Password must contain at least one special character'),
    confirmPassword: z.string({
      description: 'Confirm password',
      required_error: 'Confirm password is required',
    }),
    email: z
      .string({
        description: 'Email',
      })
      .email('Email must be a valid email'),
  })
  .refine((data) => data.password === data.confirmPassword, {
    message: 'Passwords do not match',
    path: ['confirmPassword'],
  });

export type AppSignupCredentials = z.infer<typeof appSignupCredentialsSchema>;
