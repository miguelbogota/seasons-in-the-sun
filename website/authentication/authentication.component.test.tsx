import { render } from '@testing-library/react';

import { Authentication } from './authentication.component';

describe('Website / Authentication / Component', () => {
  beforeAll(() => {
    global.GUN = () => ({
      // @ts-expect-error Mocking get method.
      get: () => ({}),
      user: () => ({
        // @ts-expect-error Mocking get method.
        recall: () => ({}),
      }),
      on: () => ({}),
    });
  });

  it('should match snapshot', () => {
    const { container } = render(<Authentication />);
    expect(container).toMatchSnapshot();
  });
});
