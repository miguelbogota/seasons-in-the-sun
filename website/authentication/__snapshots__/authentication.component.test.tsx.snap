// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Website / Authentication / Component > should match snapshot 1`] = `
<div>
  <div
    class="authentication-container"
  >
    <pre>
      null
    </pre>
    <div>
      <h2>
        Welcome Back
      </h2>
      <form>
        <div
          class="text-field"
        >
          <input
            id=":r0:"
            name="username"
            placeholder="Ex. johndoe"
            type="text"
          />
          <label
            for=":r0:"
          >
            Username
          </label>
        </div>
        <div
          class="text-field"
        >
          <input
            id=":r1:"
            name="password"
            placeholder=""
            type="text"
          />
          <label
            for=":r1:"
          >
            Password
          </label>
        </div>
        <button
          type="submit"
        >
          Sign in
        </button>
      </form>
    </div>
    <br />
    <div>
      ---------------------------------------------------------------
    </div>
    <br />
    <div>
      <h2>
        Join
      </h2>
      <form>
        <div
          class="text-field"
        >
          <input
            id=":r2:"
            name="username"
            placeholder="Ex. johndoe"
            type="text"
          />
          <label
            for=":r2:"
          >
            Username
          </label>
        </div>
        <div
          class="text-field"
        >
          <input
            id=":r3:"
            name="password"
            placeholder=""
            type="text"
          />
          <label
            for=":r3:"
          >
            Password
          </label>
        </div>
        <div
          class="text-field"
        >
          <input
            id=":r4:"
            name="confirmPassword"
            placeholder=""
            type="text"
          />
          <label
            for=":r4:"
          >
            Confirm Password
          </label>
        </div>
        <div
          class="text-field"
        >
          <input
            id=":r5:"
            name="email"
            placeholder="Ex. johndoe@email.com"
            type="text"
          />
          <label
            for=":r5:"
          >
            Email
          </label>
        </div>
        <button
          type="submit"
        >
          Sign up
        </button>
      </form>
    </div>
  </div>
</div>
`;
